### PR TITLE
Pass over Chapter 7

### PR DESCRIPTION
**Pull request by pavpanchekha**
This PR does quite a lot, most importantly:

- [x] Redo the LineLayout and TextLayout code to be more in line with the other layout objects.
- [x] Implement multi-tab support (code and text included)
- [x] Remove discussion of the CSS box model

I still have a few issues with the chapter I'd like some feedback on:

- [ ] The LineLayout/TextLayout refactor is really long and gets the chapter off to a slow start.
- [ ] I'm not sure how to order clicking links, multiple tabs, and the address bar

One thing I'm considering is putting multiple tabs first, then the address bar, and then links. I think that leads to the least code changes, but I'm not sure how to motivate multiple tabs before we have anything else. Those tabs would all view the same web page, what's the point?

### CODE REVIEW

**Comment by pavpanchekha** [comment]
@chrishtr I think the most important thing to do in your review is to look over the code, specifically, the `LineLayout`, `TextLayout`, `Tab`, and `Browser` classes. Any changes you want to make to these classes or their interfaces we should do now.

File: book/chrome.md [hunk 1]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+        and obj.y <= y < obj.y + obj.height]
+if not objs: return
+```
+
+Now, normally you click on lots of layout objects at once: some text,
+and also the paragraph it's in, and the section that that paragraph is
+in, and so on. We want the one that's "on top", so the last object in
+the list:[^overlap]
+
+[^overlap]: In a real browser you can also have a dialog that overlaps
```
Comment (chrishtr) [inline book/chrome.md:312]:
```suggestion
[^overlap]: In a real browser you can also have sibling elements that overlap each other, such as a dialog that overlaps
```

File: book/chrome.md [hunk 2]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+some text, or something like that. And real browsers can control which
```
Comment (chrishtr) [inline book/chrome.md:313]:
```suggestion
some text. And real browsers can control which
```

File: book/chrome.md [hunk 3]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+element is on top using the `z-index` property. So real browsers use
+*stacking contexts* to resolve what exactly you actually clicked on.
```
Comment (chrishtr) [inline book/chrome.md:315]:
```suggestion
[*stacking contexts*](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) to resolve what exactly you actually clicked on.
```

File: book/chrome.md [hunk 4]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+while elt:
+    if isinstance(elt, Text):
+        pass
+    elif elt.tag == "a" and "href" in elt.attributes:
+        # ...
+    elt = elt.parent
+```
+
+I wrote this in a kind of curious way, but this way it's easier to add
+other types of clickable things later---like text boxes and buttons in
```
Comment (chrishtr) [inline book/chrome.md:338]:
```suggestion
other types of clickable things later---like text boxes and buttons---in
```

File: book/chrome.md [hunk 5]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+the [next chapter](forms.md). Finally, now that we've found the link
+element itself, we need to extract the URL and direct the browser to
+it. This URL might be a relative URL:
+
+``` {.python indent=12}
+# ...
+elif elt.tag == "a" and "href" in elt.attributes:
+    url = relative_url(elt.attributes["href"], self.url)
```
Comment (chrishtr) [inline book/chrome.md:346]: Nit: relative_url is should be called absolute_url. I'll write a PR for this.

File: book/chrome.md [hunk 6]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+
+Fundamentally tabbed browsing means we need to distinguish between the
+browser itself and individual tabs that browse some specific web page.
+Right now the `Browser` class stores both information about the
+browser (like the canvas it draws to) and information about a single
+tab (like the layout tree and display list). We need to tease the two
+apart.
+
+Here's the plan: the `Browser` class will store the window and canvas,
+plus the list of tabs. Everything else goes into a new `Page` class.
```
Comment (chrishtr) [inline book/chrome.md:379]:
```suggestion
plus a list of `Tab` objects, one for each browser tab. Everything else goes into a new `Page` class.
```

File: book/chrome.md [hunk 7]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+Since the `Browser` stores the window and canvas, it handles all of
+the events, sometimes forwarding it to the active tab.
+
+So the `Tab` constructor looks like this:
```
Comment (chrishtr) [inline book/chrome.md:383]:
```suggestion
The `Tab` constructor looks like this:
```

File: book/chrome.md [hunk 8]
```diff
@@ -8,531 +8,797 @@ next: forms
...
-now don't need to do anything but recurse on their children.
+The `load`, `scrolldown`, `click`, and `draw` methods all move to
+`Tab`, since that's now where all web-page-specific data lives. But
+there's a change here too.
 
-Navigating between pages
-========================
+Since the `Browser` controls the canvas and handles events, it decides
+when rendering happens and which tab does the drawing. After all, you
+only want one tab drawing its contents at a time! So let's remove the
```
Comment (chrishtr) [inline book/chrome.md:398]:
```suggestion
only want one tab drawing its contents at a time![^windowed] So let's remove the
```

Comment (chrishtr) [reply book/chrome.md:398]: Plus:

[^windowed] Unless the browser implements multiple windows!

And add an exercise for this.

File: book/chrome.md [hunk 9]
```diff
@@ -8,531 +8,797 @@ next: forms
...
-    # ...
+    def handle_click(self, e):
+        self.tabs[self.active_tab].click(e.x, e.y)
+        self.draw()
 ```
 
-Now, inside `handle_click`, we can convert the link we clicked on to a
-new URL:
+You'll need to modify the arguments `Tab`'s `scrolldown` and `click`
+methods expect. Finally, that `draw` call also just forwards to the
```
Comment (chrishtr) [inline book/chrome.md:447]: Spell this out?

File: book/chrome.md [hunk 10]
```diff
@@ -8,531 +8,797 @@ next: forms
...
 
-[^10]: Yep, that predates and inspired the name of Google's Chrome
+[^chrome]: Yep, that predates and inspired the name of Google's Chrome
     browser.
 
-To do that, we first have to move the page content itself further down.
-I'm going to reserve 60 pixels for the browser chrome, which we need
-to subtract in `draw`:
+So let's add some code to draw a set of tabs at the top of the browser
+window. Let's try to keep it simple, because this is going to require
```
Comment (chrishtr) [inline book/chrome.md:491]:
```suggestion
window (and keep it simple, because this is going to require
```

File: book/chrome.md [hunk 11]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+some tedious and mildly tricky geometry. Let's have each tab be 80
```
Comment (chrishtr) [inline book/chrome.md:492]:
```suggestion
some tedious and mildly tricky geometry). Let's have each tab be 80
```

File: book/chrome.md [hunk 12]
```diff
@@ -8,531 +8,797 @@ next: forms
...
+them independently, and switch tabs whenever you want.
+
+Navigation history
+==================
+
+Now that we are navigating between pages all the time, it's easy to
+get a little lost and forget what web page you're looking at.
+Browsers solve this issue with an address bar that shows the URL.
+Let's implement a little address bar ourselves. The URL it shows is
+the URL for the currently-active tab:
```
Comment (chrishtr) [inline book/chrome.md:628]:
```suggestion
the one for the currently-active tab:
```

File: src/lab7.py [hunk 13]
```diff
@@ -57,550 +57,618 @@ def request(url):
 
     return headers, body
 
+def relative_url(url, current):
```
Comment (chrishtr) [inline src/lab7.py:60]: absolute_url

Comment (pavpanchekha) [reply src/lab7.py:60]: You're right. Perhaps `resolve_url`? This is going to need a PR because this function was introduced in the last chapter.

**Pull request merged**